core/src/repair/serve_repair.rs

```diff
@@ -1127,10 +1127,7 @@ impl ServeRepair {
         if repair_peers.is_empty() {
             return Err(ClusterInfoError::NoPeers.into());
         }
-        let (weights, index): (Vec<_>, Vec<_>) = cluster_slots
-            .compute_weights_exclude_nonfrozen(slot, &repair_peers)
-            .into_iter()
-            .unzip();
+        let (weights, index) = cluster_slots.compute_weights_exclude_nonfrozen(slot, &repair_peers);
```

Comment:
`WeightedShuffle::new()` only requires an `IntoIterator` for `weights` so this does not require us to `collect()` into a vector. However, we do `index[i]` a few lines down which does require a `Vec` and not an iterator.

Since `weights` and `index` are created by iterating the same item in `compute_weights_exclude_nonfrozen()`, it doesn't seem super easy to decouple these and get one iterator / one `Vec` out.

Just mentioning incase you have any clever ideas. Otherwise, I think this PR is still an improvement as we got from 3 vectors (one intermediate + two final) to 2 vectors

Reply:
the original code also did multiple reallocations :)

Reply:
> the original code also did multiple reallocations :)

Oh I agree. As I said,
> Otherwise, I think this PR is still an improvement as we got from 3 vectors (one intermediate + two final) to 2 vectors

I'm onboard with the change, was just thinking out loud to see if we could do any better (which does not appear to be the case without a larger rework)